zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1217 @@
 
-    use color_eyre::eyre::{bail, eyre, Report};
-    use tower::{util::ServiceExt, Service};
+    use color_eyre::eyre::{eyre, Report};
+    use std::{cmp::min, mem::drop, time::Duration};
+    use tokio::time::timeout;
+    use tower::{Service, ServiceExt};
 
     use zebra_chain::serialization::ZcashDeserialize;
 
+    /// The timeout we apply to each verify future during testing.
+    ///
+    /// The checkpoint verifier uses `tokio::sync::oneshot` channels as futures.
+    /// If the verifier doesn't send a message on the channel, any tests that
+    /// await the channel future will hang.
+    ///
+    /// This value is set to a large value, to avoid spurious failures due to
+    /// high system load.
+    const VERIFY_TIMEOUT_SECONDS: u64 = 10;
+
     #[tokio::test]
     #[spandoc::spandoc]
-    async fn checkpoint_single_item_list() -> Result<(), Report> {
-        install_tracing();
+    async fn single_item_checkpoint_list() -> Result<(), Report> {
+        zebra_test::init();
 
         let block0 =
             Arc::<Block>::zcash_deserialize(&zebra_test::vectors::BLOCK_MAINNET_GENESIS_BYTES[..])?;
         let hash0: BlockHeaderHash = block0.as_ref().into();
 
         // Make a checkpoint list containing only the genesis block
-        let genesis_checkpoint_list: HashMap<BlockHeight, BlockHeaderHash> =
+        let genesis_checkpoint_list: BTreeMap<BlockHeight, BlockHeaderHash> =
             [(block0.coinbase_height().unwrap(), hash0)]
                 .iter()
                 .cloned()
                 .collect();
 
-        let mut state_service = Box::new(zebra_state::in_memory::init());
-        let mut checkpoint_verifier = super::init(state_service.clone(), genesis_checkpoint_list);
+        let mut checkpoint_verifier =
+            CheckpointVerifier::new(genesis_checkpoint_list).map_err(|e| eyre!(e))?;
+
+        assert_eq!(
+            checkpoint_verifier.previous_checkpoint_height(),
+            BeforeGenesis
+        );
+        assert_eq!(
+            checkpoint_verifier.target_checkpoint_height(),
+            WaitingForBlocks
+        );
+        assert_eq!(checkpoint_verifier.max_checkpoint_height(), BlockHeight(0));
 
         /// Make sure the verifier service is ready
         let ready_verifier_service = checkpoint_verifier
             .ready_and()
             .await
             .map_err(|e| eyre!(e))?;
-        /// Verify block 0
-        let verify_response = ready_verifier_service
-            .call(block0.clone())
+        /// Set up the future for block 0
+        let verify_future = timeout(
+            Duration::from_secs(VERIFY_TIMEOUT_SECONDS),
+            ready_verifier_service.call(block0.clone()),
+        );
+        /// Wait for the response for block 0
+        // TODO(teor || jlusby): check error kind
+        let verify_response = verify_future
             .await
+            .expect("timeout should not happen")
             .map_err(|e| eyre!(e))?;
```

Comment:
So I'm working on fixing this but for now spandoc is restricted on how thoroughly it can instrument futures. The problem is it can only instrument the futures themselves, and not the rest of the expressions in the chain of function calls. The way you have this setup (and the way I setup the examples that you based this on) is subtly wrong. Right here you're constructing a `Report` in a combinator on the `Result` returned from the future, so this gets called after the future is done executing and the span is no longer entered. We need to start using `TryFutureExt::map_err` before the await to have the conversion run as part of the future if we want to get these spandoc comments in our error messages.

So if you notice your spandoc spans aren't showing up in your eyre::Reports that's why.

Reply:
Fixed in fec6d2ba47be0fffa9e8c11687a7b94a83136000